### PR TITLE
Add constructors from scalar numpy arrays

### LINKED ISSUE #1974: Inconsistant behaviour when constructing string variables
These lines execute successfully and result in `Variable`s with a dtype of `sc.dtype.string`:
```python
sc.Variable(values=np.array(["1", "2"]), dims=["x"])
sc.Variable(values=np.array(["1", "2"]), dims=["x"], dtype=sc.dtype.string)
sc.Variable(value="some_string")
```

These lines result in "unsupported dtype" errors:
```python
sc.Variable(value=np.array("1"))
sc.Variable(value=np.array("1"), dtype=sc.dtype.string)
sc.Variable(value="some_string", dtype=sc.dtype.string)
```

Equivalents of the failing lines for numeric types work fine, for example
```python
sc.Variable(value=np.array(1))
sc.Variable(value=1, dtype=sc.dtype.int32)
```

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #1974

Note that higher dimensional arrays still don't work:
```.py
sc.Variable(values=np.array([["1", "2"]]), dims=["x", "y"])
```
This would require a more complicated change away from `std::vector`.

### CODE REVIEW

**Comment by matthew-d-jones** [comment]
This fixes the situation when using a scalar array and/or specifying string type, but breaks the following case which was previously working
image

File: python/variable.cpp [hunk 1]
```diff
@@ -23,34 +23,52 @@
...
+    throw std::invalid_argument("Cannot convert values of type " +
+                                to_string(data_dtype) + " to " +
+                                to_string(requested_dtype));
+  }
+}
+} // namespace
+
 template <class T> void bind_init_0D(py::class_<Variable> &c) {
+  // Conversions are in general not possible. We have the dtype
+  // arguments only for symmetry with other constructors.
```
Comment (SimonHeybrock) [inline python/variable.cpp:54]: Can you elaborate? Can't we convert, say, `int` to `double`?

**Comment by jl-wynen** [comment]
> but breaks the following case which was previously working

Should be fixed now

File: python/variable.cpp (hunk 1 above)
Comment (jl-wynen) [reply python/variable.cpp:57]: Yes. But they are handled by `bind_init_0D_native_python_types`. This function is for cases where we cannot convert (so easily / reasonable).

**Comment by matthew-d-jones** [comment]
> > but breaks the following case which was previously working
> 
> Should be fixed now

Thanks, working well now. But should these cases be covered by unit tests?

**Pull request closed without merging**